RenderListBox::setScrollTop should allow out-of-range values but clamp them

nt1m opened the pull request:
#### cb92b7a07300a8d1e890190ee82ad00de879c2d6
<pre>
RenderListBox::setScrollTop should allow out-of-range values but clamp them
<a href="https://bugs.webkit.org/show_bug.cgi?id=248327">https://bugs.webkit.org/show_bug.cgi?id=248327</a>
rdar://102656180

Reviewed by Cameron McCormack.

The CSSOM View spec allows out-of-range values when scrolling, but it requires clamping:
<a href="https://w3c.github.io/csswg-drafts/cssom-view/#scroll-an-element">https://w3c.github.io/csswg-drafts/cssom-view/#scroll-an-element</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-multiple-scrolling.optional-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-size-scrolling-and-sizing.optional-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::setScrollTop):

Canonical link: <a href="https://commits.webkit.org/257012@main">https://commits.webkit.org/257012@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e67631430a701bf038745d579c9e3fc10acde6ec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97611 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6860 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30779 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107107 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167370 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101562 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7213 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35621 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90004 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103766 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103259 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84241 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32406 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87272 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89070 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75328 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/858 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/845 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21999 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5665 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2385 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2098 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41391 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/rendering/RenderListBox.cpp:
```diff
@@ -805,13 +805,13 @@ static void setupWheelEventTestMonitor(RenderListBox& renderer)
 
 void RenderListBox::setScrollTop(int newTop, const ScrollPositionChangeOptions&)
 {
```
Do you have a bug on file for making RenderListBox::setScrollLeft do something sensible for vertical writing modes?

nt1m replied:
That's https://bugs.webkit.org/show_bug.cgi?id=248262

heycam commented on Source/WebCore/rendering/RenderListBox.cpp:
```diff
@@ -805,13 +805,13 @@ static void setupWheelEventTestMonitor(RenderListBox& renderer)
 
     setupWheelEventTestMonitor(*this);
-    scrollToOffsetWithoutAnimation(ScrollbarOrientation::Vertical, index);
+    scrollToOffsetWithoutAnimation(ScrollbarOrientation::Vertical, std::clamp(index, 0, numItems() - 1));
```
I think we should continue to return early (and not call either of these two functions) if after clamping `index == m_indexOffset`.

Also, will this do the right thing if `numItems() == 0`? `std::clamp(0, 0, -1) == -1`. Although maybe we'll have returned early above in this case?

nt1m replied:
Hmm yeah good catch. I'll fix this.

nt1m committed:
RenderListBox::setScrollTop should allow out-of-range values but clamp them
https://bugs.webkit.org/show_bug.cgi?id=248327
rdar://102656180

Reviewed by Cameron McCormack.

The CSSOM View spec allows out-of-range values when scrolling, but it requires clamping:
https://w3c.github.io/csswg-drafts/cssom-view/#scroll-an-element

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-multiple-scrolling.optional-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-size-scrolling-and-sizing.optional-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::setScrollTop):

Canonical link: https://commits.webkit.org/257012@main
